contracts/lease/src/api/open/mod.rs

```diff
@@ -112,23 +117,111 @@ impl InterestPaymentSpec {
     }
 }
 
-#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, JsonSchema)]
+#[derive(Serialize, Clone, PartialEq, Eq, JsonSchema)]
 #[cfg_attr(any(test, feature = "testing"), derive(Debug))]
 #[serde(deny_unknown_fields, rename_all = "snake_case")]
 pub struct PositionSpec {
+    /// Liability constraints
     pub liability: Liability,
+    /// The minimum amount to liquidate. Any attempt to liquidate a smaller
+    /// amount would be postponed until the amount goes above this limit
     pub min_asset: LpnCoin,
+    ///  The minimum amount that a lease asset should be evaluated past any
+    ///  partial liquidation or close. If not, a full liquidation is performed
     pub min_sell_asset: LpnCoin,
 }
 
 impl PositionSpec {
-    #[cfg(any(test, feature = "testing"))]
     pub fn new(liability: Liability, min_asset: LpnCoin, min_sell_asset: LpnCoin) -> Self {
-        Self {
+        let obj = Self {
             liability,
             min_asset,
             min_sell_asset,
+        };
+        debug_assert_eq!(Ok(()), obj.invariant_held());
+        obj
+    }
+
+    fn invariant_held(&self) -> ContractResult<()> {
+        Self::check(
+            !self.min_asset.is_zero(),
+            "Min asset amount should be positive",
+        )
+        .and(Self::check(
+            self.min_asset.ticker() == self.min_sell_asset.ticker(),
+            "The ticker of min asset should be the same as the ticker of min sell asset",
+        ))
+    }
+
+    fn check(invariant: bool, msg: &str) -> ContractResult<()> {
+        ContractError::broken_invariant_if::<Self>(!invariant, msg)
+    }
+}
+
+impl<'de> Deserialize<'de> for PositionSpec {
```

Comment:
a far simpler solution would be to use the already proven technique `#[serde(try_from...)]`. Why have we implemented our own deserializer?